rcl/src/rcl/expand_topic_name.c

```diff
@@ -0,0 +1,252 @@
+    RCL_SET_ERROR_MSG("node name is invalid", allocator)
+    return RCL_RET_NODE_INVALID_NAME;
+  }
+  // validate the namespace
+  rmw_ret = rmw_validate_namespace(node_namespace, &validation_result, NULL);
+  if (rmw_ret != RMW_RET_OK) {
+    RCL_SET_ERROR_MSG(rmw_get_error_string_safe(), allocator)
+    switch (rmw_ret) {
+      case RMW_RET_INVALID_ARGUMENT:
+        return RCL_RET_INVALID_ARGUMENT;
+      case RMW_RET_ERROR:
+      // fall through on purpose
+      default:
+        return RCL_RET_ERROR;
+    }
+  }
+  if (validation_result != RMW_NODE_NAME_VALID) {
+    RCL_SET_ERROR_MSG("node namespace is invalid", allocator)
+    return RCL_RET_NODE_INVALID_NAMESPACE;
+  }
+  // check if the topic has substitutions to be made
+  bool has_a_substitution = strchr(input_topic_name, '{') != NULL;
+  bool has_a_namespace_tilde = input_topic_name[0] == '~';
+  bool is_absolute = input_topic_name[0] == '/';
+  // if absolute and doesn't have any substitution
+  if (is_absolute && !has_a_substitution) {
+    // nothing to do, duplicate and return
+    *output_topic_name = rcutils_strdup(input_topic_name, allocator);
+    if (!*output_topic_name) {
+      *output_topic_name = NULL;
+      RCL_SET_ERROR_MSG("failed to allocate memory for output topic", allocator)
+      return RCL_RET_BAD_ALLOC;
+    }
+    return RCL_RET_OK;
+  }
+  char * local_output = NULL;
+  // if has_a_namespace_tilde, replace that first
+  if (has_a_namespace_tilde) {
+    // special case where node_namespace is just '/'
+    // then no additional separating '/' is needed
+    const char * fmt = (strlen(node_namespace) == 1) ? "%s%s%s" : "%s/%s%s";
+    local_output =
+      rcutils_format_string(allocator, fmt, node_namespace, node_name, input_topic_name + 1);
+    if (!local_output) {
+      *output_topic_name = NULL;
+      RCL_SET_ERROR_MSG("failed to allocate memory for output topic", allocator)
+      return RCL_RET_BAD_ALLOC;
+    }
+  }
+  // if it has any substitutions, replace those
+  if (has_a_substitution) {
+    // Assumptions entering this scope about the topic string:
+    //
+    // - All {} are matched and balanced
+    // - There is no nesting, i.e. {{}}
+    // - There are no empty substitution substr, i.e. '{}' versus '{something}'
+    //
+    // These assumptions are taken because this is checked in the validation function.
+    const char * current_output = (local_output) ? local_output : input_topic_name;
+    char * next_opening_brace = NULL;
+    while ((next_opening_brace = strchr(current_output, '{')) != NULL) {
+      char * next_closing_brace = strchr(current_output, '}');
+      // conclusion based on above assumptions: next_closing_brace - next_opening_brace > 1
+      size_t substitution_substr_len = next_closing_brace - next_opening_brace + 1;
+      // figure out what the replacement is for this substitution
+      const char * replacement = NULL;
+      if (strncmp("{node}", next_opening_brace, substitution_substr_len) == 0) {
```

Comment:
i was thinking about putting these literals somewhere in constant variables, to have them exposed in a single place and can be changed solely there in case they have to be reused at some point. But just my opinion,